lib/src/rules/unnecessary_parenthesis.dart

```diff
@@ -72,6 +72,7 @@ class _Visitor extends SimpleAstVisitor<void> {
       if (parent is ConditionalExpression) return;
       if (parent is CascadeExpression) return;
       if (parent is FunctionExpressionInvocation) return;
+      if (parent is PrefixExpression) return;
```

Comment:
This seems too permissive. It allows anything to be inside the parens as long as they are preceded by a prefix operator. As an extreme example, consider `!(x)`. It seems like users would want to know about these parens. Both of the test cases you included have a space in them, which is the case listed in the issue. But do we really want to allow every nested expression?

Reply:
That extreme example would be reported by the `if (node.expression is SimpleIdentifier)` check.

But I know what you mean. There's not an easy list of nodes that always/usually have whitespace is there? That's the motivating factor, is to allow people to write "unnecessary" parens in locations where whitespace makes it look like `!` binds tighter than it does. We could make the checks something like:

```dart
if ((parent is PrefixExpression) && (
    node is AsExpression ||
    node is AwaitExpression ||
    node is InstanceCreationExpression ||
    node is IsExpression ||
    node is ListLiteral ||
    node is MapLiteral))
```